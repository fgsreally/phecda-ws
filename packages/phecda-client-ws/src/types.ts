import type { Construct, PickFunc } from 'phecda-server'

type AnyFunction = (...args: any) => any
type ParseInstance<Instance extends Record<string, AnyFunction>> = {
  [Key in keyof Instance]: ToClientFn<Instance[Key]>
}

export type ToClientMap<ControllerMap extends Record<string, Construct>> = {
  [Key in keyof ControllerMap]: ToClientInstance<InstanceType<ControllerMap[Key]>>
}

 type ToClientInstance<Instance extends Record<string, any>> = ParseInstance<PickFunc<Instance>>

 type ToClientFn<Func extends AnyFunction> = (...p: Parameters<Func>) => void
