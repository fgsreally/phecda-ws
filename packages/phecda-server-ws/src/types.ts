import type { BaseContext, DefaultOptions } from 'phecda-server'
import type WS from 'ws'
import type { WebSocket } from 'ws'
declare module 'phecda-server'{
  interface ControllerMetaData {
    ws?: { on?: boolean }
  }
}

export interface ClientEvents {
  [key: string]: any
}

export interface WsContext extends BaseContext {
  type: 'ws'
  send<Event extends keyof ClientEvents>(event: Event, data: ClientEvents[Event]): Promise<void>
  broadcast<Event extends keyof ClientEvents>(event: Event, data: ClientEvents[Event], includeCurrentConnect?: boolean): Promise<void>
  app: WS.Server
  websocket: WS
  args: any[]
}

export interface WsOptions extends DefaultOptions {
  //  false that it won't send data to browser
  isEventSentToClient?: <Key extends keyof ClientEvents>(ws: WebSocket, event: Key, data: ClientEvents[Key]) => Promise<boolean> | boolean
}
