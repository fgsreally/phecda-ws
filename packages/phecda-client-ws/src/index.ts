export function createClient(ws:WebSocket){

}