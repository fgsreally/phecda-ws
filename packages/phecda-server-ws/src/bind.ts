import type WS from 'ws'
import { Context, HMR, detectAopDep } from 'phecda-server'
import type { ControllerMeta, Factory } from 'phecda-server'
import Debug from 'debug'
import WebSocket from 'ws'
import type { ClientEvents, WsContext, WsOptions } from './types'
const debug = Debug('phecda-server/ws')

export function bind(wss: WS.Server, data: Awaited<ReturnType<typeof Factory>>, opts: WsOptions = {}) {
  const { globalGuards, globalInterceptors, globalFilter, globalPipe, isEventSentToClient = () => true } = opts
  const { moduleMap, meta } = data

  const metaMap = new Map<string, Record<string, ControllerMeta>>()
  function handleMeta() {
    metaMap.clear()
    for (const item of meta) {
      const { tag, func, controller, ws } = item.data
      if (controller !== 'ws' || !ws.on)
        continue

      debug(`register method "${func}" in module "${tag}"`)

      if (metaMap.has(tag))
        metaMap.get(tag)![func] = item as ControllerMeta

      else
        metaMap.set(tag, { [func]: item as ControllerMeta })
    }
  }

  wss.on('connection', (ws) => {
    ws.on('message', async (raw) => {
      try {
        const data = JSON.parse(raw.toString())
        const { _ps, tag, func, args } = data
        if (_ps !== 1)
          return
        const meta = metaMap.get(tag)![func]
        const contextData = {
          type: 'ws' as const,
          meta,
          moduleMap,
          tag,
          func,
          app: wss,
          websocket: ws,
          args,
          async send<Event extends keyof ClientEvents>(event: Event, data: ClientEvents[Event]) {
            if (await isEventSentToClient?.(ws, event, data))
              ws.send(JSON.stringify({ event, data }))
          },
          async broadcast<Event extends keyof ClientEvents>(event: Event, data: ClientEvents[Event], includeCurrentConnect = false) {
            const message = JSON.stringify({ event, data })

            await Promise.all([...wss.clients].map(async (socket) => {
              if (await isEventSentToClient?.(socket, event, data)) {
                if ((includeCurrentConnect || ws !== socket) && socket.readyState === WebSocket.OPEN)
                  socket.send(message)
              }
            }))
          },
        }
        const context = new Context<WsContext>(contextData)

        await context.run({ globalGuards, globalInterceptors, globalFilter, globalPipe }, returnData => returnData, (err) => {
          wss.emit('error', err)
        })
      }
      catch (e) {
      }
    })
  })

  detectAopDep(meta, {
    guards: globalGuards,
    interceptors: globalInterceptors,
  }, 'ws')
  handleMeta()

  HMR(async () => {
    detectAopDep(meta, {
      guards: globalGuards,
      interceptors: globalInterceptors,
    }, 'ws')
    handleMeta()
  })
}
