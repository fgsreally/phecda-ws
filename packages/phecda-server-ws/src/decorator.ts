import { setMeta } from 'phecda-server'

export function Ws() {
  return (target: any) => {
    setMeta(target, undefined, undefined, {
      controller: 'ws',
    })
  }
}

export function On(target: any, property: PropertyKey) {
  setMeta(target, property, undefined, {
    ws: {
      on: true,
    },
  })
}
