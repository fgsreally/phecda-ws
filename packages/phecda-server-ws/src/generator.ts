import type { ControllerMetaData, MetaData } from 'phecda-server'
import { Generator } from 'phecda-server'

export class WSGenerator extends Generator {
  name = 'WS'

  classMap: Record<string, { [key: string]: string }> = {}

  getContent() {
    let content = ''

    for (const name in this.classMap) {
      content += `
        export class ${name}{
            ${Object.values(this.classMap[name]).reduce((p, c) => p + c)}
            }`
    }
    return content
  }

  addMethod(args: ControllerMetaData) {
    const {
      ws, name, func, tag,
    } = args
    if (!ws?.on)
      return
    if (!this.classMap[name])
      this.classMap[name] = {}
    this.classMap[name][func] = `
    ${func}(...args){
return {tag:"${tag}",func:"${func}",args}
    }
    `
  }

  generateCode(meta: MetaData[]): string {
    for (const i of meta) {
      if (i.controller === 'ws')
        this.addMethod(i as ControllerMetaData)
    }
    return this.getContent()
  }
}
