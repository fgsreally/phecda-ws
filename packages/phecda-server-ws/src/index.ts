import type { ServerOptions } from 'ws'
import Websocker from 'ws'
import { Init, SHARE_KEY, getExposeKey, getOwnState, getState, getTag, setState, setStateKey } from 'phecda-server'
type WrapServer<Server extends Websocker.Server, Events extends Record<string, any>> = Server & {
  send<Key extends keyof Events>(event: Key, data: Events[Key]): void
}

export class WS<Events extends Record<string, any>> {
  server: Websocker.Server
  constructor(options?: ServerOptions) {
    // @ts-expect-error just type magic
    this.server = new Websocker.Server(options)
  }

  send() {

  }

  @Init
  init() {
    this.server.on('connection', (socket) => {
      const keys = getExposeKey(this).filter(item => typeof item === 'string') as string[]
      const eventRecord: Record<string, string> = {}
      for (const key of keys) {
        const { ws } = getState(this, key)
        if (ws)
          eventRecord[ws] = key
      }

      socket.on('message', (raw) => {
        const data = JSON.parse(raw.toString())
        this[eventRecord[data.type]]!(data.data)
      })
    })
  }
}

export function On(event?: string) {
  return (target: any, key: string) => {
    let state: any
    if (!key) {
      if (!event)
        event = getTag(target)
      key = SHARE_KEY

      target = target.prototype
      const state = getOwnState(target, key) || {}
      state.ws = event
    }
    else {
      if (!event)
        event = key

      const prefix = getOwnState(target, SHARE_KEY)?.ws
      const state = getOwnState(target, key) || {}
      state.ws = prefix ? `${prefix}-${event}` : event

      if (!state.define)
        state.define = {}

      state.define.ws = state.ws
    }

    setState(target, key, state)
    setStateKey(target, key)
  }
}
