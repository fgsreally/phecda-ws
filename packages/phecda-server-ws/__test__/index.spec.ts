import { expect, test } from 'vitest'
import { Factory } from 'phecda-server'
import { WebSocketServer } from 'ws'
import { WSGenerator, bind } from '../src'
import { TestWs } from './fixture/test.ws'

test('WSGenerator', async () => {
  const data = await Factory([TestWs])
  const code = new WSGenerator('').generateCode(data.meta.map(item => item.data))

  expect(code).toMatchSnapshot()
})

test('meta', async () => {
  const data = await Factory([TestWs])
  expect(data.meta).toMatchSnapshot()
})
test('meta when bind', async () => {
  const data = await Factory([TestWs])
  bind(new WebSocketServer({ noServer: true }), data, { globalGuards: ['a'], globalInterceptors: ['a'] })
  expect(data.meta).toMatchSnapshot()
})
