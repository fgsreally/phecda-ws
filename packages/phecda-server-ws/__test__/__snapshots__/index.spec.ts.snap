// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`WSGenerator 1`] = `
"
        export class TestWs{
            
    add(...args){
return {tag:"TestWs",func:"add",args}
    }
    
            }"
`;

exports[`meta 1`] = `
[
  Meta {
    "data": {
      "controller": "ws",
      "ctx": "context",
      "define": {},
      "filter": undefined,
      "func": "add",
      "guards": [],
      "interceptors": [],
      "name": "TestWs",
      "params": [
        {
          "define": {},
          "index": 0,
          "key": "0",
          "pipe": undefined,
          "type": "args",
        },
      ],
      "plugins": [],
      "tag": "TestWs",
      "ws": {
        "on": true,
      },
    },
    "paramsType": [
      [Function],
    ],
  },
]
`;

exports[`meta when bind 1`] = `
[
  Meta {
    "data": {
      "controller": "ws",
      "ctx": "context",
      "define": {},
      "filter": undefined,
      "func": "add",
      "guards": [],
      "interceptors": [],
      "name": "TestWs",
      "params": [
        {
          "define": {},
          "index": 0,
          "key": "0",
          "pipe": undefined,
          "type": "args",
        },
      ],
      "plugins": [],
      "tag": "TestWs",
      "ws": {
        "on": true,
      },
    },
    "paramsType": [
      [Function],
    ],
  },
]
`;
